[#40546] Global roles for groups

cbliard opened the pull request:
https://community.openproject.org/work_packages/40546

cbliard committed:
add global role to users when added to a group having global role

cbliard committed:
Filter out global roles when editing project memberships for group

cbliard committed:
Update global role for users belonging to a group having global role

cbliard committed:
Cleanup global role associated with group when membership is deleted

It was already handled. Only tests were updated.

cbliard committed:
fix rubocop warning

RSpec/ExampleWording: Do not use should when describing your tests.

cbliard committed:
fix indentation

cbliard committed:
Make rubocop happier

cbliard committed:
Add 'Global role' tab for groups

Global roles can now be added, update, and removed for a group

cbliard committed:
Harmonize params name for membership creation and update

This will allow further factorization

cbliard committed:
make rubocop happier

Rails/MatchRoute: use the http verb instead of match
Rails/HelperInstanceVariable: do not use instance variables in view
helpers

cbliard committed:
Fix misplaced end statement

cbliard committed:
merge some common global roles files from users and groups

cbliard committed:
Avoid instance variable in view helpers

rubocop rule Rails/HelperInstanceVariable

cbliard committed:
merge common global roles view files from users and groups

cbliard committed:
hide deletion button if global role is inherited from group

cbliard committed:
rubocop: add 'as' as appropriate prefix for RSpec/ContextWording

cbliard committed:
send emails when adding global role to a group

Also refactor the logic to decide if a 'membership created' or 'membership updated' email should be sent

cbliard committed:
display membership emails in the language of the recipient

cbliard committed:
Make rubocop happier

Remove the 'FactoryBot.'

cbliard committed:
fix test

cbliard committed:
add api memberships tests related to global role

also adjust Member visibility so that a group without any members can be queried

cbliard committed:
Make rubocop happy

cbliard committed:
fix bug: no more anonymous user when notifying of group membership removals

before, the email sent on group membership removal/ deletion was from Anonymous user. Now it is from the real user having performed the action in the administration

cbliard committed:
rewrite test in a better way

cbliard committed:
add cbliard in the admins list for pullpreview

cbliard committed:
update docs about gobal roles for groups

ulferts commented on app/mailers/member_mailer.rb:
```diff
@@ -78,8 +78,8 @@ def alter_project(current_user, member, subject, message)
-    in_member_locale(member) do
-      User.execute_as(current_user) do
+    User.execute_as(current_user) do
+      in_member_locale(member) do
```
:+1: thanks

ulferts commented on app/workers/mails/member_created_job.rb:
```diff
@@ -32,19 +32,11 @@ class Mails::MemberCreatedJob < Mails::MemberJob
-    if new_roles_added?(user_member, group_member)
-      send_updated_project(current_user, user_member, message)
-    elsif all_roles_added?(user_member, group_member)
+    difference = Mails::MemberRolesDiff.new(user_member, group_member)
```
Nitpicking here but I'd place it in `Members::RolesDiff` (app -> models -> members) since it's applicability is not only limited to mails. It compares two members based on the roles they have.

cbliard replied:
Good point. Done in ed3aa3f598.

cbliard committed:
Merge branch 'dev' into feature/40546-global-roles-for-groups

ulferts reviewed (commented):
This is not a full review but so far, the changes are looking good. Up until https://github.com/opf/openproject/pull/10063/files#diff-a38463e91a7b872b60542b658f0123dc3d7a63bd284287fb20f45235616b15be I only had the tiny remark. I did not try out the change yet.

oliverguenther commented on app/models/member.rb:
```diff
@@ -70,6 +70,10 @@ def deletable?
   end
 
+  def deletable_role?(role)
+    member_roles.filter { _1.role_id == role.id }.none?(&:inherited_from)
```
I can see that this doesn't make a lot of difference in the roles table where all member roles are loaded anyway, but I'd prefer a query like the following to avoid doing the filter in ruby.

```suggestion
    member_roles
      .where(role_id: role.id)
      .where.not(inherited_from: nil)
      .empty?
```

cbliard replied:
Good point. Done in b6e4dbcbcc.

I introduced a scope for better readibility.

oliverguenther reviewed (approved):
Functionality-wise, LGTM. I have that one small performance related remark. If you can address that and the one from Jens, good to merge 👏

cbliard committed:
prefer db query over pure ruby for #deletable_role?

cbliard committed:
rename Mails::MemberRolesDiff as Members::RolesDiff

applicability is not only limited to emails, hence mving it to models makes more sense.

oliverguenther commented:
:+1: Feel free to merge once you checked the documentation part you talked about

cbliard committed:
document how to get global memberships in API
